spanner/retry.go

```diff
@@ -57,6 +58,16 @@ func onCodes(bo gax.Backoff, cc ...codes.Code) gax.Retryer {
 // Retry returns the retry delay returned by Cloud Spanner if that is present.
 // Otherwise it returns the retry delay calculated by the generic gax Retryer.
 func (r *spannerRetryer) Retry(err error) (time.Duration, bool) {
+	if status.Code(err) == codes.Internal &&
```

Comment:
This will cause `SpannerRetryer` to always retry `Internal` errors with these specific error messages, regardless whether the user has specified `Internal` as an error code that should be retried or not. That means that:
* Other `Internal` errors will never be retried, even if the user has specified `Internal` as a retryable error code.
* These specific `Internal` errors will always be retried, even when the user has not specified `Internal` as a retryable code.

I would suggest that we move this check to what is now line 75 (i.e. after the check whether this error code should be retried at all). I also think we should consider adding an additional configuration parameter to `SpannerRetryer` that specifies whether all `Internal` errors should be retried, or only these specific errors to avoid any confusion about what is being retried and what is not.

Reply:
@olavloite Please take a look at the code again. `This will cause SpannerRetryer to always retry Internal errors with these specific error messages` => This is not true. The following code only says that if it is an internal error and its error message is not in the list, then return `false`. If it is an internal error and it error message in the list, then let it pass to the next check `r.Retryer.Retry(err)`. 

This is to filter non-retryable internal errors.

Reply:
In terms of adding an option to retry all internal errors or specific ones, I wonder when should we retry all internal errors? Do we need this option? 

The new introduced option would be a bit strange: 

```go
type spannerRetryer struct {
	gax.Retryer
        retryAllInternalErrors bool
}

func onCodes(bo gax.Backoff, retryAllInternalErrors bool, cc ...codes.Code) gax.Retryer {
	return &spannerRetryer{
		Retryer: gax.OnCodes(cc, bo),
                 retryAllInternalErrors: retryAllInternalErrors
	}
}
```

Do you have any better idea?

Reply:
> This is not true. The following code only says that:
> 
> * if it is an internal error and its error message is not in the list, then return `false`.
> * If it is an internal error and it error message in the list, then let it pass to the next check `r.Retryer.Retry(err)`.

You're completely right. Sorry, I read that wrong.

Reply:
> In terms of adding an option to retry all internal errors or specific ones, I wonder when should we retry all internal errors? Do we need this option?

No, we currently don't need the option for retrying all internal errors. I'm only worried about confusion for someone reading the calling code which says `retryer := onCodes(d.backoff, codes.Unavailable, codes.Internal)`, or someone using the `spannerRetryer` for some other method where he/she also wants to retry `Internal` errors.

There's no clue here that the retryer will only retry some specific `Internal` errors, which could lead to confusion if a different `Internal` error were to be thrown by a backend method, which would then not be retried.

So I would prefer an additional method that does what your suggested solution above does, but with a description of what it's for, so something like this in addition to the currently existing `onCodes` method:

```go

// Retries all errors with one of the codes in cc, and all errors with code Internal that have
// one of a list of known descriptions that should be retried.
func onCodesAndKnownInternalErrors(bo gax.Backoff, cc ...codes.Code) gax.Retryer {
	return &spannerRetryer{
		Retryer: gax.OnCodes(cc, bo),
                retryKnownInternalErrors: true,
	}
}
```

Another option might be to just add some additional comments to the current `onCodes` method to indicate what happens when you add `Internal` to the list of codes. It is an internal struct and method, and I don't feel very strongly either way, so I'll leave it up to you what you think is best.

Reply:
Thanks for the suggestion. I prefer the second one (adding some additional comments) because both `onCodesAndKnownInternalErrors` and `retryKnownInternalErrors` confuses me that it must retry known internal errors, however, it does not.